src/symbol/collision_index.js

```diff
@@ -84,56 +87,34 @@ class CollisionIndex {
         };
     }
 
-    approximateTileDistance(tileDistance: any, lastSegmentAngle: number, pixelsToTileUnits: number, cameraToAnchorDistance: number, pitchWithMap: boolean): number {
-        // This is a quick and dirty solution for chosing which collision circles to use (since collision circles are
-        // laid out in tile units). Ideally, I think we should generate collision circles on the fly in viewport coordinates
-        // at the time we do collision detection.
-        // See https://github.com/mapbox/mapbox-gl-js/issues/5474
-
-        // incidenceStretch is the ratio of how much y space a label takes up on a tile while drawn perpendicular to the viewport vs
-        //  how much space it would take up if it were drawn flat on the tile
-        // Using law of sines, camera_to_anchor/sin(ground_angle) = camera_to_center/sin(incidence_angle)
-        // Incidence angle 90 -> head on, sin(incidence_angle) = 1, no stretch
-        // Incidence angle 1 -> very oblique, sin(incidence_angle) =~ 0, lots of stretch
-        // ground_angle = u_pitch + PI/2 -> sin(ground_angle) = cos(u_pitch)
-        // incidenceStretch = 1 / sin(incidenceAngle)
-
-        const incidenceStretch = pitchWithMap ? 1 : cameraToAnchorDistance / this.pitchfactor;
-        const lastSegmentTile = tileDistance.lastSegmentViewportDistance * pixelsToTileUnits;
-        return tileDistance.prevTileDistance +
-            lastSegmentTile +
-            (incidenceStretch - 1) * lastSegmentTile * Math.abs(Math.sin(lastSegmentAngle));
-    }
-
-    placeCollisionCircles(collisionCircles: Array<number>,
-                          allowOverlap: boolean,
-                          scale: number,
-                          textPixelRatio: number,
+    placeCollisionCircles(allowOverlap: boolean,
                           symbol: any,
                           lineVertexArray: SymbolLineVertexArray,
                           glyphOffsetArray: GlyphOffsetArray,
                           fontSize: number,
                           posMatrix: mat4,
                           labelPlaneMatrix: mat4,
+                          labelToScreenMatrix?: mat4,
                           showCollisionCircles: boolean,
                           pitchWithMap: boolean,
-                          collisionGroupPredicate?: any): { circles: Array<number>, offscreen: boolean } {
+                          collisionGroupPredicate?: any,
+                          circlePixelDiameter: number,
+                          textPixelPadding: number): { circles: Array<number>, offscreen: boolean, collisionDetected: boolean } {
         const placedCollisionCircles = [];
 
-        const projectedAnchor = this.projectAnchor(posMatrix, symbol.anchorX, symbol.anchorY);
+        const perspectiveRatio = this.projectAnchor(posMatrix, symbol.anchorX, symbol.anchorY).perspectiveRatio;
```

Comment:
Both `projectAnchor(...)` and `projection.project(...)` (a couple lines later) project the point. Maybe `projectAnchor` can be removed completely? It looks like the perspectiveRatio math was already previously duplicated here, in draw_symbol.js and in projection.js. Maybe all three of those can be deduplicated as well?

Reply:
That's a great idea! I've added a function to projection.js for computing perspective ratio. Two projections is still required in collision_index though.

Reply:
Ahh, right. I didn't notice they were being projected to different planes.